presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduNodePartitioningProvider.java

```diff
@@ -0,0 +1,93 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.kudu;
+
+import io.prestosql.spi.connector.BucketFunction;
+import io.prestosql.spi.connector.ConnectorBucketNodeMap;
+import io.prestosql.spi.connector.ConnectorNodePartitioningProvider;
+import io.prestosql.spi.connector.ConnectorPartitionHandle;
+import io.prestosql.spi.connector.ConnectorPartitioningHandle;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.connector.ConnectorSplit;
+import io.prestosql.spi.connector.ConnectorTransactionHandle;
+import io.prestosql.spi.connector.SchemaTableName;
+import io.prestosql.spi.type.Type;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.function.ToIntFunction;
+import java.util.stream.IntStream;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.prestosql.spi.connector.ConnectorBucketNodeMap.createBucketNodeMap;
+import static java.util.Objects.requireNonNull;
+
+public class KuduNodePartitioningProvider
+        implements ConnectorNodePartitioningProvider
+{
+    private final KuduClientSession clientSession;
+
+    @Inject
+    public KuduNodePartitioningProvider(KuduClientSession clientSession)
+    {
+        this.clientSession = requireNonNull(clientSession, "clientSession is null");
+    }
+
+    @Override
+    public List<ConnectorPartitionHandle> listPartitionHandles(
+            ConnectorTransactionHandle transactionHandle,
+            ConnectorSession session,
+            ConnectorPartitioningHandle partitioningHandle)
+    {
+        KuduPartitioningHandle handle = (KuduPartitioningHandle) partitioningHandle;
+        return IntStream.range(0, handle.getBucketCount())
+                .mapToObj(KuduPartitionHandle::new)
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public ConnectorBucketNodeMap getBucketNodeMap(
+            ConnectorTransactionHandle transactionHandle,
+            ConnectorSession session,
+            ConnectorPartitioningHandle partitioningHandle)
+    {
+        KuduPartitioningHandle handle = (KuduPartitioningHandle) partitioningHandle;
+        return createBucketNodeMap(handle.getBucketCount());
+    }
+
+    @Override
+    public ToIntFunction<ConnectorSplit> getSplitBucketFunction(
+            ConnectorTransactionHandle transactionHandle,
+            ConnectorSession session,
+            ConnectorPartitioningHandle partitioningHandle)
+    {
+        return value -> ((KuduSplit) value).getBucketNumber();
+    }
+
+    @Override
+    public BucketFunction getBucketFunction(
+            ConnectorTransactionHandle transactionHandle,
+            ConnectorSession session,
+            ConnectorPartitioningHandle partitioningHandle,
+            List<Type> partitionChannelTypes,
+            int bucketCount)
```

Comment:
Any reason `bucketCount` gets ignored here?

Reply:
Well, `bucketCount` means the total count of bucket, and I can get more informations about bucket schema in attribute `KuduBucketFunction#hashBucketSchemas`, so I didn't use this param.   :)